app/multitenant/dynamo_collector.go

```diff
@@ -345,6 +408,72 @@ func (c *dynamoDBCollector) getReports(userid string, row int64, start, end time
 	return append(cachedReports, fetchedReport...), nil
 }
 
+func memcacheStatusCode(err error) string {
+	// See https://godoc.org/github.com/bradfitz/gomemcache/memcache#pkg-variables
+	switch err {
+	case nil:
+		return "200"
+	case memcache.ErrCacheMiss:
+		return "404"
+	case memcache.ErrMalformedKey:
+		return "400"
+	default:
+		return "500"
+	}
+}
+
+func (c *dynamoDBCollector) fetchFromMemcache(reportKeys []string) ([]report.Report, []string, error) {
+	var foundReports map[string]*memcache.Item
+	err := timeRequest("Get", memcacheRequestDuration, memcacheStatusCode, func() error {
+		var err error
+		foundReports, err = c.memcache.GetMulti(reportKeys)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, reportKeys, err
+	}
+
+	// Decode all the reports in parallel.
+	type result struct {
+		key    string
+		report *report.Report
+	}
+	ch := make(chan result, len(reportKeys))
+	var missing []string
+	for _, reportKey := range reportKeys {
+		item, ok := foundReports[reportKey]
+		if ok {
+			go func(reportKey string) {
+				rep, err := report.MakeFromBinary(bytes.NewReader(item.Value))
+				if err != nil {
+					log.Warningf("Corrupt report in memcache %v: %v", reportKey, err)
+					ch <- result{key: reportKey}
+				} else {
+					ch <- result{report: rep}
+					c.cache.Set(reportKey, *rep)
```

Comment:
I feel this cache set should be moved to the outer function, although I'm not sure.